hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -403,6 +426,81 @@ protected R processExpiredEntry(Data key, R record, long expiryTime, long now, S
+        if (cacheConfig.getBackupCount() == 0) {
+            return;
+        }
+        if (key != null && record != null) {
+            this.expiredKeys.offer(new ExpiredKey(toHeapData(key), record.getCreationTime()));
+        }
+
+        sendBackupExpirations(true);
+    }
+
+    @Override
+    public boolean isExpirable() {
+        return hasEntryWithExpiration || getConfig().getExpiryPolicyFactory() != null;
+    }
+
+    @SuppressWarnings({"checkstyle:npathcomplexity"})
+    public void sendBackupExpirations(boolean checkIfReachedBatch) {
+        InvalidationQueue<ExpiredKey> invalidationQueue = this.expiredKeys;
+
+        int size = invalidationQueue.size();
+        if (size == 0 || checkIfReachedBatch && size < MAX_EXPIRED_KEY_COUNT_IN_BATCH) {
+            return;
+        }
+
+        if (!invalidationQueue.tryAcquire()) {
+            return;
+        }
+
+        Collection<ExpiredKey> expiredKeys = collectExpiredKeys(invalidationQueue);
+        if (expiredKeys.size() == 0) {
+            return;
+        }
+
+        // send expired keys to all backups
+        OperationService operationService = nodeEngine.getOperationService();
+        int backupReplicaCount = cacheConfig.getTotalBackupCount();
+        for (int replicaIndex = 1; replicaIndex < backupReplicaCount + 1; replicaIndex++) {
+            if (canSendBackupExpiration(getPartitionId(), replicaIndex)) {
+                Operation operation = new CacheExpireBatchBackupOperation(getName(), expiredKeys, size());
+                operationService.createInvocationBuilder(CacheService.SERVICE_NAME, operation, getPartitionId())
+                        .setReplicaIndex(replicaIndex).invoke();
+            }
+        }
+    }
+
+    private Collection<ExpiredKey> collectExpiredKeys(InvalidationQueue<ExpiredKey> invalidationQueue) {
+        Collection<ExpiredKey> expiredKeys = new ArrayList<ExpiredKey>();
+        try {
+            ExpiredKey key;
+            do {
+                key = invalidationQueue.poll();
+                if (key != null) {
+                    expiredKeys.add(key);
+                }
+            } while (key != null);
+        } finally {
+            invalidationQueue.release();
+        }
+        return expiredKeys;
+    }
+
+    protected boolean canSendBackupExpiration(int partitionId, int replicaIndex) {
+        Address replicaAddress = nodeEngine.getPartitionService().getPartition(partitionId).getReplicaAddress(replicaIndex);
+        if (replicaAddress == null) {
+            return false;
+        }
+        // Previous versions did not remove expired entries until they are touched. Old members behave the same whereas
+        // newer members still benefit from periodic removal of expired entries.
+        //
+        // RU_COMPAT_3_10
+        return nodeEngine.getClusterService().getMember(replicaAddress).getVersion().asVersion().isGreaterOrEqual(Versions.V3_11);
```

Comment:
This looks good to me, can you verify that the cluster version can be compared like this? @vbekiaris

Reply:
Yes, that's the way to ensure a remote member's codebase version is at least 3.11.x. I am however concerned about the potential for a race: between the time `canSendBackupExpiration` determines the owner for a partiticular `{partitionId, replicaIndex}` and the time this is re-evaluated in `invoke()`, a topology change can result in sending out a `CacheExpireBatchBackupOperation` to a 3.10.x member, leading to serialization exception on the 3.10.x member. I see two possible mitigations:
- don't send out backup expirations unless cluster version is >= 3.11
- invoke the operation on the specific target member that was already identified in `canSendBackupExpiration`

wdyt?

Reply:
@vbekiaris I changed the code. It now sends the operation to already identified member.